### PR TITLE
tsdb/wal: refactoring error messages for list segments.

### PR DESCRIPTION
**Pull request by johncming**
https://github.com/prometheus/prometheus/blob/2255b6f62f4aed34a62cb4bf2ad645e72bbba037/tsdb/wal/wal.go#L767

`listSegments` is used in multiple places with inconsistent error messages. It was refactored so that it could maintain a more consistent output makes it easier for users to spot this error.

Use private scopes to keep changes to a minimum.

### CODE REVIEW

File: tsdb/wal/wal.go [hunk 1]
```diff
@@ -105,6 +105,15 @@ func (e *CorruptionErr) Error() string {
 	return fmt.Sprintf("corruption in segment %s at %d: %s", SegmentName(e.Dir, e.Segment), e.Offset, e.Err)
 }
 
+type listSegmentsErr struct {
```
Comment (codesome) [inline tsdb/wal/wal.go:108]: I don't think there is any benefit of an explicit error type here. We can include the directory error message at those 3 places in listSegments.

Comment (johncming) [reply tsdb/wal/wal.go:108]: Make sense.

File: tsdb/wal/wal.go [hunk 2]
```diff
@@ -767,7 +767,7 @@ type segmentRef struct {
 func listSegments(dir string) (refs []segmentRef, err error) {
 	files, err := ioutil.ReadDir(dir)
 	if err != nil {
-		return nil, err
+		return nil, fmt.Errorf("list segments in dir %s: %s", dir, err)
```
Comment (Harkishen-Singh) [inline tsdb/wal/wal.go:770]: Any reasons why we are not using `errors.Wrapf()`?

File: tsdb/wal/wal.go [hunk 3]
```diff
@@ -782,7 +782,7 @@ func listSegments(dir string) (refs []segmentRef, err error) {
 	})
 	for i := 0; i < len(refs)-1; i++ {
 		if refs[i].index+1 != refs[i+1].index {
-			return nil, errors.New("segments are not sequential")
+			return nil, fmt.Errorf("list segments in dir %s: %s", dir, errors.New("segments are not sequential"))
```
Comment (Harkishen-Singh) [inline tsdb/wal/wal.go:785]: IMO `errors.Wrapf()` should be preferred.

File: tsdb/wal/wal.go (hunk 2 above)
Comment (johncming) [reply tsdb/wal/wal.go:770]: thanks, pkg `errors` is better.

File: tsdb/wal/wal.go [hunk 4]
```diff
@@ -782,7 +782,7 @@ func listSegments(dir string) (refs []segmentRef, err error) {
 	})
 	for i := 0; i < len(refs)-1; i++ {
 		if refs[i].index+1 != refs[i+1].index {
-			return nil, errors.New("segments are not sequential")
+			return nil, errors.Errorf("list segments in dir %s: %s", dir, errors.New("segments are not sequential"))
```
Comment (codesome) [inline tsdb/wal/wal.go:785]:
```suggestion
			return nil, errors.Errorf("list segments in dir %s: %s", dir, "segments are not sequential")
```

**Review by codesome — APPROVED** [review]

File: tsdb/wal/wal.go (hunk 4 above)
Comment (codesome) [reply tsdb/wal/wal.go:785]: Poking @johncming in case you missed this. Let's get this merged!

Comment (bwplotka) [reply tsdb/wal/wal.go:785]: ping (:

**Comment by roidelapluie** [comment]
We have looked at this pull request during our bug scrub.

cc @johncming @jesusvazquez, is this pull request still relevant? Are you willing to rebase it?

Thank you for your contribution.

**Comment by beorn7** [comment]
:wave: from the bug scrub. We still think this is a good idea, but no rebase happened, and by now, we have moved to a different error package. Let's close this, but whoever wants to pick this up, please re-create with the currently used error package.

**Pull request closed without merging**